libgalois/include/katana/PropertyGraph.h

```diff
@@ -767,6 +819,620 @@ CreateSymmetricGraph(PropertyGraph* pg);
+  EdgeLabelIndex(EdgeLabelIndex&&) = default;
+  EdgeLabelIndex& operator=(EdgeLabelIndex&&) = default;
+
+  EdgeLabelIndex(const EdgeLabelIndex&) = delete;
+  EdgeLabelIndex& operator=(const EdgeLabelIndex&) = delete;
+
+  static EdgeLabelIndex Make(const PropertyGraph* pg) noexcept;
+
+  EdgeLabelTy label(uint32_t index) const noexcept {
+    KATANA_LOG_ASSERT(index < num_unique_edge_labels_);
+    KATANA_LOG_ASSERT(size_t(index) < edge_index_to_label_map_.size());
+    return edge_index_to_label_map_[index];
+  }
+
+  uint32_t index(const EdgeLabelTy& label) const noexcept {
+    KATANA_LOG_ASSERT(edge_label_to_index_map_.count(label) > 0);
+    // return edge_label_to_index_map_[label];
+    return edge_label_to_index_map_.find(label)->second;
+  }
+
+  uint32_t num_unique_labels() const noexcept {
+    return num_unique_edge_labels_;
+  }
+
+  /**
+   * @param label: label to check
+   * @returns true iff there exists some edge in the graph with that label
+   */
+  bool has_edge_label(const EdgeLabelTy& label) const noexcept {
+    return (
+        edge_label_to_index_map_.find(label) !=
+        edge_label_to_index_map_.cend());
+  }
+#if 0
+  /**
+   * Wrapper to get the begin iterator to distinct edge labels in the graph.
+   *
+   * @returns Iterator to first distinct edge label
+   */
+  auto DistinctEdgeLabelsBegin() const noexcept {
+    return edge_index_to_label_map_.cbegin();
+  }
+
+  /**
+   * Wrapper to get the end iterator to distinct edge labels in the graph.
+   *
+   * @returns Iterator to end distinct edge label
+   */
+  auto DistinctEdgeLabelsEnd() const noexcept {
+    return edge_index_to_label_map_.cend();
+  }
+#endif
+
+  /**
+   * Wrapper to get the distinct edge labels in the graph.
+   *
+   * @returns Range of the distinct edge labels
+   */
+  auto distinct_edge_labels() const noexcept {
+    return MakeStandardRange(
+        edge_index_to_label_map_.cbegin(), edge_index_to_label_map_.cend());
+  }
+
+private:
+  // map an integer id to each unique edge label in the graph
+  using EdgeLabelToIndexMap = std::unordered_map<EdgeLabelTy, uint32_t>;
+  // each index refers to a unique edge label in the graph
+  using EdgeIndexToLabelMap = std::vector<EdgeLabelTy>;
+
+  EdgeLabelIndex(
+      EdgeLabelToIndexMap&& edge_label_to_index,
+      EdgeIndexToLabelMap&& edge_index_to_label, uint32_t num_labels) noexcept
+      : edge_label_to_index_map_(std::move(edge_label_to_index)),
+        edge_index_to_label_map_(std::move(edge_index_to_label)),
+        num_unique_edge_labels_(num_labels) {
+    KATANA_LOG_DEBUG_ASSERT(
+        num_labels == edge_index_to_label_map_.size() &&
+        num_labels == edge_label_to_index_map_.size());
+  }
+
+  EdgeLabelToIndexMap edge_label_to_index_map_;
+  EdgeIndexToLabelMap edge_index_to_label_map_;
+  uint32_t num_unique_edge_labels_ = 0u;
+};
+
+class KATANA_EXPORT EdgeLabelAwareTopology;
+
+// TODO(amber): make OrigEdgeIDMap optional via template argument
```

Comment:
I want to be really careful with this. Configuration via template is very hard to wrap from Python. That said if there are a few common choices and the objects are always passed around via pointer or reference (using dynamic typing) then it would be OK.

Reply:
see above.